### PR TITLE
Several optimization to improve memory usage

### PR DESCRIPTION
**Pull request by ryzhyk**
See commit messages for details

### CODE REVIEW

File: src/Language/DifferentialDatalog/Compile.hs [hunk 1]
```diff
@@ -1943,6 +2043,71 @@ mkType' TOpaque{..}                = rname typeName <>
...
+typeSize' :: DatalogProgram -> Type -> Int
+typeSize' _ TBool{}    = 1
+typeSize' _ TInt{}     = 32
+typeSize' _ TString{}  = 24
+typeSize' _ TBit{..} | typeWidth <= 8   = 1
+                     | typeWidth <= 16  = 2
+                     | typeWidth <= 32  = 4
+                     | typeWidth <= 64  = 8
+                     | typeWidth <= 128 = 16
+                     | otherwise        = 32
```
Comment (mbudiu-vmw) [inline src/Language/DifferentialDatalog/Compile.hs]: not something depending on the bitvector size?

File: src/Language/DifferentialDatalog/Compile.hs [hunk 2]
```diff
@@ -1943,6 +2043,71 @@ mkType' TOpaque{..}                = rname typeName <>
...
+typeSize' d (TStruct _ [cons]) = consSize d $ consArgs cons
+typeSize' d (TStruct _ cs) =
+    tag_size + (maximum $ 0 : map (consSize d . consArgs) cs)
+    where
+    tag_align = maximum $ 1 : map (consAlignment d . consArgs) cs
+    tag_size = pad 4 tag_align
+typeSize' d (TTuple _ as) = tupleSize d as
+typeSize' _ TOpaque{}     = 0xffffffff -- be conservative
```
Comment (mbudiu-vmw) [inline src/Language/DifferentialDatalog/Compile.hs]: this looks very conservative

Comment (mbudiu-vmw) [reply src/Language/DifferentialDatalog/Compile.hs]: Maybe opaque types have to implement some interface to estimate the size?

Comment (mbudiu-vmw) [reply src/Language/DifferentialDatalog/Compile.hs]: Or maybe they know when to box themselves?

File: src/Language/DifferentialDatalog/DatalogProgram.hs [hunk 3]
```diff
@@ -144,19 +163,41 @@ progAtomMapM d fun = do
...
+depNodeIsRel _              = False
+
+type DepGraph = G.Gr DepGraphNode Bool
 
 -- | Dependency graph among program relations.  An edge from Rel1 to
 -- Rel2 means that there is a rule with Rel1 in the right-hand-side,
 -- and Rel2 in the left-hand-side.  Edge label is equal to the
 -- polarity with which Rel1 occurs in the rule.
 --
+-- In addition, we conservitively add both a positive and a negative edge
```
Comment (mbudiu-vmw) [inline src/Language/DifferentialDatalog/DatalogProgram.hs]: typo

File: src/Language/DifferentialDatalog/DatalogProgram.hs [hunk 4]
```diff
@@ -144,19 +163,41 @@ progAtomMapM d fun = do
...
+-- from Rel1 to Rel2 if they appear respectively as input and output of a
+-- transformer application (since we currently don't have a way of knowing
+-- if the transformer).
```
Comment (mbudiu-vmw) [inline src/Language/DifferentialDatalog/DatalogProgram.hs]: unfinished phrase. This does not really look conservative to me.

File: src/Language/DifferentialDatalog/Validate.hs [hunk 5]
```diff
@@ -314,6 +319,93 @@ ruleCheckAggregate d rl idx = do
...
+                      HOTypeFunction{} -> check (not $ isUpper $ head $ name o) (pos o) "Function name may not start with an upper-case letter"
+          ) $ transInputs ++ transOutputs
+    mapM_ (hotypeValidate d . hofType) $ transInputs ++ transOutputs
+
+-- | Validate transformer application
+-- * Transformer exists
+-- * Inputs and outputs refer to valid functions and relations
+-- * Input/output types match transformer declaration
+-- * Outputs cannot be bound to input relations
+-- * Outputs of a transformer cannot be used in the head of a rule
```
Comment (mbudiu-vmw) [inline src/Language/DifferentialDatalog/Validate.hs]: I thought you wanted to allow this. But it's fine this way.

File: src/Language/DifferentialDatalog/Validate.hs [hunk 6]
```diff
@@ -314,6 +319,93 @@ ruleCheckAggregate d rl idx = do
...
+    bindings <- unifyTypes d (pos a) ("in transformer application " ++ show a) $ concat types
+    mapM_ (\ta -> case M.lookup ta bindings of
+                       Nothing -> err (pos a) $ "Unable to bind type argument '" ++ ta ++
+                                                " to a concrete type in transformer application " ++ show a
+                       Just t  -> return ())
+          $ transformerTypeVars trans
+    mapM_ (\o -> check (relRole (getRelation d o) /= RelInput) (pos a)
+                 $ "Transformer output cannot be bound to input relation " ++ o
+          ) applyOutputs
+    -- Things will break if a relation is assigned by an 'Apply' in the top scope and then occurs inside a
```
Comment (mbudiu-vmw) [inline src/Language/DifferentialDatalog/Validate.hs]: I see.

File: src/Language/DifferentialDatalog/Validate.hs [hunk 7]
```diff
@@ -314,6 +319,93 @@ ruleCheckAggregate d rl idx = do
...
+          applyOutputs
+    -- Likewise, to relax this, modify 'compileApplyNode' to concatenate transformer output to
+    -- existing relation if it exists.
+    mapM_ (\o -> check (length (relApplys d o) == 1) (pos a)
+                       $ "Relation " ++ o ++ " occurs as output of multiple transformer applications")
+          applyOutputs
+
+hotypeValidate :: (MonadError String me) => DatalogProgram -> HOType -> me ()
+hotypeValidate d HOTypeFunction{..} = do
+    -- FIXME: hacky way to validate function type by converting it into a function.
```
Comment (mbudiu-vmw) [inline src/Language/DifferentialDatalog/Validate.hs]: why is it hacky?

File: rust/template/ddlog.h [hunk 8]
```diff
@@ -233,6 +233,9 @@ extern int ddlog_clear_relation(ddlog_prog prog, table_id table);
  * Dump the content of an output table by invoking `cb` for each value
  * in the table.
  *
+ * `cb` returns `true` to allow enumeration to continue or `false` to
```
[the remark this answers is not in the document]
Comment (ryzhyk) [reply rust/template/ddlog.h]: I did not change anything; just clarified the semantics in the comment

File: rust/template/differential_datalog/program.rs [hunk 9]
```diff
@@ -102,15 +159,23 @@ pub struct Program<V: Val> {
...
+
+/// Program node is either an individual non-recursive relation, a transformer application or
 /// a vector of one or more mutually recursive relations.
 #[derive(Clone)]
 pub enum ProgNode<V: Val> {
-    RelNode{rel: Relation<V>},
-    SCCNode{rels: Vec<Relation<V>>}
+    Rel{rel: Relation<V>},
+    Apply{tfun: TransformerFunc<V>},
+    SCC{rels: Vec<Relation<V>>}
```
[the remark this answers is not in the document]
Comment (ryzhyk) [reply rust/template/differential_datalog/program.rs]: No, this is completely unrelated to the SCC transformer.
The compiler builds a dependency graph of relations and extracts SCCs from it, which correspond to recursive fragments that live inside their own nested scopes in DD.

File: src/Language/DifferentialDatalog/Validate.hs (hunk 7 above)
Comment (ryzhyk) [reply src/Language/DifferentialDatalog/Validate.hs]: A clean solution is to change the function type to basically be a named variable of higher-order type.

**Pull request merged**